keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1699 @@
++       // +optional
++       Container []AllowedQoSResource
++}
+
++// AllowedQoSResource specifies access to one QoS-class resources type.
++type AllowedQoSResource struct {
++       // Name of the resource.
++       Name QoSResourceName
++       // Allowed classes.
++       Classes []string
++       // Capacity is the hard limit for usage of the class.
++       // +optional
++       Capacity int64
++}
+
+
+ // ResourceQuotaStatus defines the enforced hard limits and observed use.
+ type ResourceQuotaStatus struct {
+        ...
+        // Used is the current observed total usage of the resource in the namespace
+        // +optional
+        Used ResourceList
++       // QoSResources contains the enforced set of available class resources.
++       // +featureGate=QoSResources
++       // +optional
++       QoSResources QoSResourceQuota
+ }
+
+
+```
+## Proposal
+
+This section currently covers [implementation phase 1](#phase-1) (see
+[implementation phases](#implementation-phases) for an outline of the complete
+implementation).
+
+We extend the CRI protocol to contain information about the QoS-class
+resource assignment of containers and pods. Resource assignment requests will
+be simple key-value pairs (*resource-type=class-name*)
+
+We also extend the CRI protocol to support updates of QoS-class resource
+assignment of running containers. We recognize that currently container
+runtimes lack the capability to update either of the two types of QoS-class
+resources we have identified (RDT and blockio). However, there is no technical
+limitation in that and we are planning to implement update support for them
+in the future.
+
+We also extend the Kubernetes API to support the assignment of QoS-class
+resources to applications via PodSpec, and to users and the kube-scheduler to
+see availability of QoS-class resources on a per-node basis via NodeStatus.
+
+Container runtime is expected to be aware of all resource types and the classes
+within. The CRI protocol is extended to be able to communicate the available
+QoS-class resources from the runtime to the client. This information includes:
+- Available QoS-class resource types.
+- Available classes within each resource type.
+- Whether the resource type is immutable or if it supports in-place updates.
+  In-place updates of resoures might not be possible because of runtime
+  limitations or the underlying technology, for example.
+
+QoS-class resources may be bounded in the way that the number of applications
+that can be assigned to a specific class (of one QoS-class resource) on a node
+can be limited. This limit is configuratble on a per-class (and per-node)
+basis. This can be used to e.g. limit access to a high-tier class.
+
+Pod-level and container-level QoS-class resources are completely independent
+resource types. E.g. specifying something in the pod-level request does not
+mean specifying a pod-level default for all containers of the pod.
```

Comment:
I could think of e.g. basically any cgroup-related controls (not that it would currently be neatly possible to control pod cgroups from the runtime level, but anyway...). You might have a pod-level QoS but you'd not necessarily enforce that as a default for all containers that don't have explicitly specified that same QoS resource.

Reply:
Most of cgroups is inherently hierarchical - when does setting a pod QoS not imply the containers underneath it?

Reply:
> Most of cgroups is inherently hierarchical - when does setting a pod QoS not imply the containers underneath it?

I think it is very possible that one wants to apply totally different parameterization on the pod-level cgroup vs. container-level cgroup. Think about prioritization for example - pod-level setting is about the pod contending against other pods but the container level setting is about contending against other containers within the pod. Or maybe you'd want to set some limits/throttling on the pod level but configure only priority on the container level.

But now that I've thought about this subject once again I suddenly came up with one possible idea regarding pod vs. container level QoS resources and defaults:
1. We would deny registering pod and container level QoS resource on the same name (which might be wise in any case)
2. In addition, we would allow specifying a container-level QoS resource in the pod-level resource field in the PodSpec
3. This (container-level QoS resources specified in pod-level field) would be regarded as a default for containers of the pod. 
4. Pod-level QoS-resources could not be specified in the container level, however.

Any thoughts on this @thockin @sftim?

Reply:
sounds plausible.

How do the failures get surfaced?  It seems like it can't be static (API time).  If I put a pod-level QoS-resource in a container's specification, does it just never schedule?

Reply:
Kubernetes has so many extension points that - IMO - it's a shame not to use them.

How I'd recommend doing it: if a QoS class assignment only works at the Pod level, use either ValidatingAdmissionPolicy or a similar mechanism to reject Pods that try to specify it at the container level. And vice versa.

The default should be to allow specifying a QoS class assignment for the Pod and then to override that for a specific container. That's the most flexible, and will fix some circumstances. For example, the fictional `kubernetes.io/container-filesystem-storage` QoS class.

Bonus points for warning on the equivalent mistake in a Pod template.

To make this more easy, we could define an **entirely optional** CRD for QoS classes and an out-of-tree controller that sets up the admission time validation. Then, in a cluster that has that CRD and controller, you define your new QoSClass and everything works great. Try to set an unrecognized QoSClass? We can fail that at admission time too.

---

However, there's more than one way to do it. I'll wager there's more than seven, even.

Reply:
> How do the failures get surfaced? It seems like it can't be static (API time). If I put a pod-level QoS-resource in a container's specification, does it just never schedule?

Mm, there was an obvious gap in my quick solution 🧐 In my reasoning I was only considering the "official" names in which case it's obvious. We know at the API level and thus also in the scheduler which are pod level qos resources and which are container level resources. We can have an admission check for that and reject. But for vendor/runtime specific "non-official" it wouldn't be possible with the current design (and the pod would just be pending) but maybe that would be enought(?)

Reply:
As long as it stays pending with a useful error that I can easily find, I guess it's as good as we can do.

Reply:
I think we can recommend that if you're adding a custom QoS class resource to a cluster, you should also implement your own admission controls so that the pending pod situation doesn't arise.

We won't lead the horses to water, but we can document that they get thirsty.

Reply:
We don't really have a notion of admission allowance, just "lack of deny".  If I fat-finger a resource name and say "example.com/f00" when I meant "example.com/foo", the validation is going to pass and the pod will pend forever.  We don't have a canonical "list of resource names known to any node in the cluster" and I am not sure we want that.

In this case, we can probaly do better than *just* pending forever - but I don't know if it's a requirement.  Admission control on pods is a dangerous game - I can see how people might not want a lot of those :)

Reply:
I agree on dangerous precedent of admission controls for vendor' reousce names... It can be implemented, but that's not cheap operation on scale, because information from all nodes needs to be collected...

That's why, I'd suggest to keep it simple: follow the same practice as with extended resources/device plugins.
That means that pods with wrong names are accepted, but will be in pending state forever.
People can use `Descheduler` https://github.com/kubernetes-sigs/descheduler#podlifetime  to clean-up those.

Reply:
I agree. We can do admission-time validation of the "official" names (at GA as it's now proposed in the KEP) but vendor-specific resources would be outside of this, also in terms of recommended practices